vm: `create_vcpu` takes `u64` argument #123

gabhijit opened the pull request:
Changed the passed parameter type from `u8` to `u64`, which the
underlying `ioctl` uses.

Added a missing cap `MaxVcpuId` (It is defined in kvm-bindings for both
arm* and x86* but was missing inside cap.rs. This was required by a test
case.)

Added a test case to assert failure if we pass parameter higher than the
`max_vcpu_id`.

Signed-off-by: Abhijit Gadgil <gabhijit@iitbombay.org>

gabhijit committed:
vm: `create_vcpu` takes `u64` argument #123

Changed the passed parameter type from `u8` to `u64`, which the
underlying `ioctl` uses.

Added a missing cap `MaxVcpuId` (It is defined in kvm-bindings for both
arm* and x86* but was missing inside cap.rs. This was required by a test
case.)

Added a test case to assert failure if we pass parameter higher than the
`max_vcpu_id`.

Signed-off-by: Abhijit Gadgil <gabhijit@iitbombay.org>

gabhijit commented:
Pretty much all the build failures have the following failures - 

```~~~ Preparing plugins
[90m$[0m cd /var/lib/buildkite-agent/builds/ip-172-31-101-33-1/rust-vmm/kvm-ioctls-ci
[31m🚨 Error: Plugins have been disabled on this agent with `--no-local-hooks`[0m
^^^ +++
```
Looks like some CI issue?

andreeaflorescu commented:
> Pretty much all the build failures have the following failures -
> 
> ```
> �[90m$�[0m cd /var/lib/buildkite-agent/builds/ip-172-31-101-33-1/rust-vmm/kvm-ioctls-ci
> �[31m🚨 Error: Plugins have been disabled on this agent with `--no-local-hooks`�[0m
> ^^^ +++
> ```
> 
> Looks like some CI issue?

Yap, thanks for reporting it. It is now fixed, I re-triggered the CI, and it looks like all tests pass now.

gabhijit commented on src/ioctls/vm.rs:
```diff
@@ -1799,6 +1799,38 @@ mod tests {
+        }
+        if max_vcpus == 0 {
+            max_vcpus = 4
+        }
```
I think, this should be made a separate private function something like `vm.get_max_vcpu_id`? This is perhaps useful outside this test case as well.

andreeaflorescu replied:
yes, I remember there was such a code somewhere, but I can't seem to remember where.

andreeaflorescu replied:
Ok, I got it now. You can use the `get_max_vcpus` function that already exists in the Kvm interface.

gabhijit replied:
`get_max_vcpus` and `get_max_vcpu_id` are different ioctls as per the [documentation here](https://www.kernel.org/doc/html/latest/virt/kvm/api.html#kvm-create-vcpu) . For instance when I checked on `x86_64` on my laptop, `max_vcpus` was 288 but `max_vcpu_id` was 1023. So I believe a function like `get_max_vcpu_id` should be added to `kvm` interface? 

and then that can be used in this test case?

andreeaflorescu replied:
Yap, that sounds like a good idea.

gabhijit committed:
system: Added an API `get_max_vcpu_id` to `kvm`

Added a new API `get_max_vcpu_id` that returns the
`KVM_CAP_MAX_VCPU_ID` if the Extension is present or `get_max_vcpus`.

Also updated the test case for `vm.create_cpu` to use `max_vcpu_id`
obtained using the above API.

Updated coverage score to 91.3 to fix build issue

Signed-off-by: Abhijit Gadgil <gabhijit@iitbombay.org>

gabhijit commented:
The build failure appears due to some `coverage` issue not sure how to fix it?

andreeaflorescu commented:
> The build failure appears due to some `coverage` issue not sure how to fix it?

Oh yeah, we did not fix the coverage error message. You need to update the [coverage configuration file](https://github.com/rust-vmm/kvm-ioctls/blob/master/coverage_config_x86_64.json) with the new coverage score.

gabhijit commented:
> > The build failure appears due to some `coverage` issue not sure how to fix it?
> 
> Oh yeah, we did not fix the coverage error message. You need to update the [coverage configuration file](https://github.com/rust-vmm/kvm-ioctls/blob/master/coverage_config_x86_64.json) with the new coverage score.

Updated. The build passes now.

andreeaflorescu commented on src/ioctls/vm.rs:
```diff
@@ -1799,6 +1799,23 @@ mod tests {
     }
 
+    #[test]
+    fn create_vcpu_different_cpuids() {
```
nit: Can you rename this to `create_vcpu_different_ids`? We're using `cpuid` for something else.

gabhijit replied:
done!

gabhijit commented:
I believe this issue can be closed, right?
